clients/src/main/resources/common/message/FetchSnapshotResponse.json

```diff
@@ -0,0 +1,59 @@
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+{
+  "apiKey": 59,
+  "type": "response",
+  "name": "FetchSnapshotResponse",
+  "validVersions": "0",
+  "flexibleVersions": "0+",
+  "fields": [
+    { "name": "ThrottleTimeMs", "type": "int32", "versions": "0+", "ignorable": true,
+      "about": "The duration in milliseconds for which the request was throttled due to a quota violation, or zero if the request did not violate any quota." },
+    { "name": "ErrorCode", "type": "int16", "versions": "0+", "ignorable": false,
+      "about": "The top level response error code." },
+    { "name": "Topics", "type": "[]TopicSnapshot", "versions": "0+",
+      "about": "The topics to fetch.", "fields": [
+      { "name": "Name", "type": "string", "versions": "0+", "entityType": "topicName",
+        "about": "The name of the topic to fetch." },
+      { "name": "Partitions", "type": "[]PartitionSnapshot", "versions": "0+",
+        "about": "The partitions to fetch.", "fields": [
+        { "name": "Index", "type": "int32", "versions": "0+",
+          "about": "The partition index." },
+        { "name": "ErrorCode", "type": "int16", "versions": "0+",
+          "about": "The error code, or 0 if there was no fetch error." },
+        { "name": "SnapshotId", "type": "SnapshotId", "versions": "0+",
+          "about": "The snapshot endOffset and epoch fetched",
+          "fields": [
+          { "name": "EndOffset", "type": "int64", "versions": "0+" },
+          { "name": "Epoch", "type": "int32", "versions": "0+" }
+        ]},
+        { "name": "CurrentLeader", "type": "LeaderIdAndEpoch",
+          "versions": "0+", "taggedVersions": "0+", "tag": 0, "fields": [
+          { "name": "LeaderId", "type": "int32", "versions": "0+",
+            "about": "The ID of the current leader or -1 if the leader is unknown."},
+          { "name": "LeaderEpoch", "type": "int32", "versions": "0+",
+            "about": "The latest known leader epoch"}
+        ]},
+        { "name": "Size", "type": "int64", "versions": "0+",
+          "about": "The total size of the snapshot." },
+        { "name": "Position", "type": "int64", "versions": "0+",
+          "about": "The starting byte position within the snapshot included in the Bytes field." },
+        { "name": "Bytes", "type": "bytes", "versions": "0+", "zeroCopy": true,
```

Comment:
Can you remind me if we are planning to change the type to "records"? I don't think we will get the benefit of `sendfile` unless we do so.

Reply:
Yeah. I created this issue: https://issues.apache.org/jira/browse/KAFKA-10694

I think we have two options:
1. Use "records" as you have suggested.
2. Similar to "records" introduce a new type called for example `Memory` that has two implementations one that encapsulates `ByteBuffer` and another that encapsulates a file and a start position. This is similar to "records", `MemoryRecords` and `FileRecords` without the requirement that the start needs to be a valid `RecordBatch`.

I want to play around with this soon and restart the conversation on that issue and PR.

Reply:
@jsancio Can we add an index file to the snapshot file so that the `FetchSnaphotResponseData` can also satisfy the requirement that the start is a valid `RecordBatch`? I can help with some work of this issue: https://issues.apache.org/jira/browse/KAFKA-10694.

Reply:
@dengziming, Thanks for the comment. The `FetchSnapshotRequest` differs from `FetchRequest` in that `Position` is a byte offset. Because of this we don't need an `Offset` to `Byte` index.

If we implement option 1 in my comment above, we can still use `Records` by making sure that:
1. The `leaders` send at least one entire `RecordBatch` in the `FetchSnapshotResponse`.
2. The `followers` only write complete `RecordBatch`es by not writing bytes sent by the leader that are not complete `RecordBatch`.

This guarantees that the `Position` sent in `FetchSnapshotRequest` is `RecordBatch` aligned and it always makes progress because the leader sends at least one complete `RecordBatch`.

Having said that, I think we should try and implement option 2. Unfortunately, the Jira https://issues.apache.org/jira/browse/KAFKA-10694 doesn't document option 2.